### PR TITLE
Implement `keysend` sending support for spontaneous payments

### PR DESCRIPTION
**Pull request by cdecker**
This has been a long time in the making, but we finally are able to build new features on top of the paymod series.

This modifier is rather simple, but that's all it takes: add the payment preimage, that we generate randomly, into the onion payload for the final hop, so a recipient that knows to look for it can claim the incoming payment.

## Todo
 - [x] Write the man-page for `lightning-cli keysend`

### CODE REVIEW

File: plugins/keysend.c [hunk 1]
```diff
@@ -2,17 +2,186 @@
...
+	size_t hopcount;
+	u8 *raw_preimage;
+
+	if (p->step == PAYMENT_STEP_GOT_ROUTE) {
+		last_hop = &p->route[tal_count(p->route) - 1];
+		if (last_hop->style == ROUTE_HOP_LEGACY)
+			payment_fail(p,
+				     "Destination does not support receiving "
+				     "keysend payments (not a TLV hop)");
+	}
```
Comment (rustyrussell) [inline plugins/keysend.c:66]: I'm not sure we should check this: we might not have a node announcement, and we might as well try it?

File: plugins/keysend.c [hunk 2]
```diff
@@ -2,17 +2,186 @@
...
+	if (p->step != PAYMENT_STEP_ONION_PAYLOAD)
+		return payment_continue(p);
+
+	raw_preimage = tal_dup_arr(p->createonion_request, u8, d->preimage.r,
+				   sizeof(d->preimage), 0);
+
+	hopcount = tal_count(p->createonion_request->hops);
+	last_payload = &p->createonion_request->hops[hopcount - 1];
+	tlvstream_set_raw(&last_payload->tlv_payload->fields, PREIMAGE_TLV_TYPE,
+			  take(raw_preimage));
```
Comment (rustyrussell) [inline plugins/keysend.c:77]: Aside: maybe tlvstream_set_raw should take const void *, size_t.  That would avoid this temp allocation.

File: plugins/keysend.c [hunk 3]
```diff
@@ -184,6 +353,8 @@ static const struct plugin_hook hooks[] = {
 	},
 };
 
+
+
 int main(int argc, char *argv[])
```
Comment (rustyrussell) [inline plugins/keysend.c:358]: Spurious whitespace change :)

File: doc/lightning-keysend.7.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+generated on the destination, and the only way sender could have it is by
+sending a payment. Please ensure that this matches your use-case when using
+`keysend`.
+
+`destination` is the 33 byte, hex-encoded, node ID of the node that the payment should go to.
+`msatoshi` is in millisatoshi precision; it can be a whole number, or a whole number with suffix `msat` or `sat`, or a three decimal point number with suffix `sat`, or an 1 to 11 decimal point number suffixed by `btc`.
+
+The `label` field is used to attach a label to payments, and is returned in lightning-listpays(7) and lightning-listsendpays(7).
+The `maxfeepercent` limits the money paid in fees, and defaults to *0.5*.
+The `maxfeepercent` is a percentage of the amount that is to be paid.
```
Comment (rustyrussell) [inline doc/lightning-keysend.7.md:30]: Combine these two sentences?

**Comment by vindard** [comment]
Out of curiosity, does this supercede #3611 or simply build on top of it?

**Pull request merged**